pkg/tsdb/loki/api.go

```diff
@@ -191,7 +193,11 @@ func (api *LokiAPI) DataQuery(ctx context.Context, query lokiQuery, responseOpts
 		err := readLokiError(resp.Body)
 		lp = append(lp, "status", "error", "error", err)
 		api.log.Error("Error received from Loki", lp...)
-		return nil, err
+		// errors should be processed by errorsource middleware
+		// here we do here something extra - turning non-200 to error
+		//  we will consider this Plugin error, but let's re-evaluate if we need this
+		// @todo Re-evaluate if we need to turn non-200 to error
+		return errorsource.Response(errorsource.PluginError(err, false))
```

Comment:
the middleware will always return an error now if statuscode >= 400,  so it seems this will never be reached.

unless the status code is >= 300 and < 400

seems here we treat 300 level status codes here ( redirect, etc ) as errors?  is that right?

Reply:
Yes exactly. I don't think this is needed, but I didn't want to remove it within this PR. Therefore I added `@todo` to re-evaluate if this even needed as other data sources are not doing it.

Reply:
Seems like you lose the `readLokiError(resp.Body)` - not sure that's important though and/or if someone relies/depends on that?

Reply:
Think I already mentioned this elsewhere. I'm pretty sure we'll not provide this middleware in the public SDK eventually since it changes the normal behavior of HTTP client by returning both response and error and in Go it's expected that either of them are set, but not both. Perfectly fine to use this now, but heads up the code might require to be changed accordingly eventually.

Reply:
> Seems like you lose the readLokiError(resp.Body) - not sure that's important though and/or if someone relies/depends on that?

We are still using it in `readLokiError(resp.Body)` in `return errorsource.Response(errorsource.PluginError(err, false))`

```
		err := readLokiError(resp.Body)
		lp = append(lp, "status", "error", "error", err)
		api.log.Error("Error received from Loki", lp...)
		// errors should be processed by errorsource middleware
		// here we do here something extra - turning non-200 to error
		//  we will consider this Plugin error, but let's re-evaluate if we need this
		// @todo Re-evaluate if we need to turn non-200 to error
		return errorsource.Response(errorsource.PluginError(err, false))
```

Reply:
> Perfectly fine to use this now, but heads up the code might require to be changed accordingly eventually.

Yes, we are absolutely 100% ready to change this as it evolves further. But we also would like to move with our SLOs and test `error_source` out. And see what the real SLIs are. :)